File: app/models/questions/EssayAnswer.java
/*
 * Copyright (c) 2018 The members of the EXAM Consortium (https://confluence.csc.fi/display/EXAM/Konsortio-organisaatio)
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package models.questions;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.OneToOne;
import models.Attachment;
import models.api.AttachmentContainer;
import models.base.OwnedModel;

@Entity
public class EssayAnswer extends OwnedModel implements AttachmentContainer {

    @Column(columnDefinition = "TEXT")
    private String answer;

    @OneToOne(cascade = CascadeType.ALL)
    protected Attachment attachment;

    @Column
    private Double evaluatedScore;

    public String getAnswer() {
        return answer;
    }

    public void setAnswer(String answer) {
        this.answer = answer;
    }

    @Override
    public Attachment getAttachment() {
        return attachment;
    }

    @Override
    public void setAttachment(Attachment attachment) {
        this.attachment = attachment;
    }

    public Double getEvaluatedScore() {
        return evaluatedScore;
    }

    public void setEvaluatedScore(Double evaluatedScore) {
        this.evaluatedScore = evaluatedScore;
    }

    public EssayAnswer copy() {
        EssayAnswer essayAnswer = new EssayAnswer();
        essayAnswer.setAnswer(answer);
        essayAnswer.save();
        if (attachment != null) {
            final Attachment copy = attachment.copy();
            copy.save();
            essayAnswer.setAttachment(copy);
        }
        return essayAnswer;
    }
}

File: app/models/questions/EssayAnswer.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package models.questions

import com.fasterxml.jackson.annotation.JsonProperty
import jakarta.persistence.*
import models.attachment.{Attachment, AttachmentContainer}
import models.base.OwnedModel

import scala.compiletime.uninitialized

@Entity
class EssayAnswer extends OwnedModel with AttachmentContainer:
  @OneToOne(cascade = Array(CascadeType.ALL)) var attachment: Attachment = uninitialized

  @JsonProperty var answer: String     = uninitialized
  var evaluatedScore: java.lang.Double = uninitialized

  def copy(): EssayAnswer =
    val ea = new EssayAnswer
    ea.answer = answer
    ea.save()
    if Option(attachment).isDefined then
      val copy = attachment.copy()
      copy.save()
      ea.attachment = copy
    ea
